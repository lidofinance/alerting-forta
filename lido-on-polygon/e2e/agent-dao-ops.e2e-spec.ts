import { configureContainer, Finding } from "forta-agent";
import { AwilixContainer, asFunction } from "awilix";
import {
  provideAgentPath,
  provideRunBlock,
  provideRunTransaction,
} from "./utils";

const TEST_TIMEOUT = 60_000; // ms

describe("agent-dao-ops e2e tests", () => {
  let runBlock: (blockHashOrNumber: string | number) => Promise<Finding[]>;
  let runTransaction: (txHash: string) => Promise<Finding[]>;
  let logSpy: jest.SpyInstance;

  beforeAll(() => {
    logSpy = jest.spyOn(console, "log");
    logSpy.mockImplementation(() => {});
  });

  beforeEach(async () => {
    const container = configureContainer() as AwilixContainer;
    container.register({
      agentPath: asFunction(provideAgentPath("agent-dao-ops")),
      runTransaction: asFunction(provideRunTransaction),
      runBlock: asFunction(provideRunBlock),
    });

    // https://docs.forta.network/en/latest/cli/#invoke-commands-programmatically
    runTransaction = container.resolve("runTransaction");
    runBlock = container.resolve("runBlock");
  });

  afterAll(() => {
    jest.resetAllMocks();
  });

  it(
    "should process tx with rewards distribution",
    async () => {
      const findings = await runTransaction(
        "0xba6222301a11142cd376ceba48b390d8df97502f776dd4dcb0f180ec87f9cd4e",
      );
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT,
  );

  it(
    "should process tx with polygon checkpoint reward change",
    async () => {
      const findings = await runTransaction(
        "0xfe9ef91c9b05aac2cdad146e90dd8145f8664408c0343504d5dc64077bf9d223",
      );
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT,
  );

  it(
    "should ignore tx with checkpoint reward change event on wrong contract",
    async () => {
      const findings = await runTransaction(
        "0x5a47f08254e7590f9b92e00a517cfaaa94742407b8208e660680de44a9e0ebfc",
      );
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT,
  );

  it(
    "should process tx with pooled MATIC delegation",
    async () => {
      const findings = await runTransaction(
        "0x95b78c32c06fda9b11facbe40316cb593020a9fee58f56158c6b7481e47afbca",
      );
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT,
  );

  it(
    "should process block with low deposit executor balance",
    async () => {
      const findings = await runBlock(17089400);
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT,
  );

  it(
    "should alert on low deposit executor balance only for 100th blocks",
    async () => {
      const findings = await runBlock(16534001);
      expect(findings.length).toEqual(0);
    },
    TEST_TIMEOUT,
  );
});
