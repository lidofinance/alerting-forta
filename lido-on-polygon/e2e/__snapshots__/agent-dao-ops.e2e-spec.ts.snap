// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`agent-dao-ops e2e tests should process block with low deposit executor balance 1`] = `
Finding {
  "addresses": [],
  "alertId": "LOW-DEPOSIT-EXECUTOR-BALANCE",
  "description": "Balance of deposit executor (0xa22d223e732a5dcf4ff4529aa9a135293b7258fe) is 0.1339 ETH. This is extremely low!",
  "labels": [],
  "metadata": {},
  "name": "⚠️ Low deposit executor balance (Lido on Polygon)",
  "protocol": "ethereum",
  "severity": 4,
  "type": 2,
}
`;

exports[`agent-dao-ops e2e tests should process tx with polygon checkpoint reward change 1`] = `
Finding {
  "addresses": [],
  "alertId": "STMATIC-CHEKPOINT-REWARD-UPDATE",
  "description": "Change 41.50%, 
new reward: 71795, 
old reward: 50739",
  "labels": [],
  "metadata": {},
  "name": "ℹ Polygon checkpoint reward was changed",
  "protocol": "ethereum",
  "severity": 4,
  "type": 2,
}
`;

exports[`agent-dao-ops e2e tests should process tx with pooled MATIC delegation 1`] = `
Finding {
  "addresses": [],
  "alertId": "STMATIC-CONTRACT-POOLED-MATIC-DELEGATED",
  "description": "Pooled MATIC was delegated to validators. Delegated amount 679.9k MATIC remained pooled 0 wei",
  "labels": [],
  "metadata": {
    "args": "679886151216827958837108,0",
  },
  "name": "✅ stMATIC: Pooled MATIC delegated",
  "protocol": "ethereum",
  "severity": 1,
  "type": 4,
}
`;

exports[`agent-dao-ops e2e tests should process tx with rewards distribution 1`] = `
Finding {
  "addresses": [],
  "alertId": "STMATIC-CONTRACT-REWARDS-DISTRIBUTED",
  "description": "Rewards for stMATIC was distributed. Rewards amount 571.50 MATIC",
  "labels": [],
  "metadata": {
    "args": "571502046258362721768",
  },
  "name": "✅ stMATIC: Rewards distributed",
  "protocol": "ethereum",
  "severity": 1,
  "type": 4,
}
`;
